lib/services/local/access_list.go

```diff
@@ -93,33 +97,46 @@ func (a *AccessListService) ListAccessLists(ctx context.Context, pageSize int, n
 
 // GetAccessList returns the specified access list resource.
 func (a *AccessListService) GetAccessList(ctx context.Context, name string) (*accesslist.AccessList, error) {
-	accessList, err := a.service.GetResource(ctx, name)
+	var accessList *accesslist.AccessList
+	err := a.service.RunWhileLocked(ctx, lockName(name), accessListLockTTL, func(ctx context.Context, _ backend.Backend) error {
+		var err error
+		accessList, err = a.service.GetResource(ctx, name)
+		return trace.Wrap(err)
+	})
 	return accessList, trace.Wrap(err)
 }
 
 // UpsertAccessList creates or updates an access list resource.
 func (a *AccessListService) UpsertAccessList(ctx context.Context, accessList *accesslist.AccessList) (*accesslist.AccessList, error) {
-	if err := a.service.UpsertResource(ctx, accessList); err != nil {
+	err := a.service.RunWhileLocked(ctx, lockName(accessList.GetName()), accessListLockTTL, func(ctx context.Context, _ backend.Backend) error {
+		return trace.Wrap(a.service.UpsertResource(ctx, accessList))
+	})
+	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 	return accessList, nil
 }
 
 // DeleteAccessList removes the specified access list resource.
 func (a *AccessListService) DeleteAccessList(ctx context.Context, name string) error {
-	// Delete all associated members.
-	err := a.DeleteAllAccessListMembersForAccessList(ctx, name)
-	if err != nil {
-		return trace.Wrap(err)
-	}
+	err := a.service.RunWhileLocked(ctx, lockName(name), accessListLockTTL, func(ctx context.Context, _ backend.Backend) error {
+		// Delete all associated members.
+		err := a.memberService.WithPrefix(name).DeleteAllResources(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		return trace.Wrap(a.service.DeleteResource(ctx, name))
+	})
 
-	return trace.Wrap(a.service.DeleteResource(ctx, name))
+	return trace.Wrap(err)
 }
 
 // DeleteAllAccessLists removes all access lists.
 func (a *AccessListService) DeleteAllAccessLists(ctx context.Context) error {
+	// Locks are not used here as these operations are more likely to be used by the cache.
```

Comment:
could you clarify a bit more why this doesn't use locks? I'm not sure how it relates to cache

Reply:
We could potentially perform locks here, but we'd need to get N locks for every access list. Unless there's some sort of wildcard lock. Thoughts?

Reply:
Talking offline with @rosstimothy, I think I'll avoid the lock here and then just have the gRPC backend return `NotImplemented` for these functions. They're really only used by the collections in `lib/cache/collections.go`, they're not expected to be called by users.

Reply:
I'm hesitant to do that for two reasons:

1. I don't know for certain that these aren't used outside of the cache, I can only assert that the access list bits are.
2. I don't want to overcomplicate this particular PR, as it's only concerned with access lists.